### ISSUE TITLE
Loading too diverse

### ISSUE THREAD

**Issue by astronouth7303** [xonsh/xonsh#1464]
There are currently 4 non-interactive ways of loading and executing code in xonsh (as of 84501037eb10d9bda963435e8a33166edb3aea01, shortly after the release of v0.4.4):

* Python import
* `xontrib load`
* `.xonshrc`
* sourcing

Currently, these all have slightly different semantics and rules. 

* Python `import` works as expected (the current virtual environment does not apply, but doing that would be a whole other can of worms)
* `source` loads up the entire file and executes it in the shell context
* `xontrib load` runs the xontrib in its own namespace, but uses slightly different importing rules than `from ... import *` (it ignores `__all__` and just copies everything not prefixed with a `_`)
* `xonshrc` is basically the same as sourcing, but it has a different code path

I propose to bring `xontrib load` in line with `import` by specifying `xontrib load` to behave the same as `from xontrib.<name> import *` (plus name resolution rules). Namely, by respecting `__all__` if the xontrib defines it. This reduces the effective number of types of executions to two (sourcing and importing, or shell-style and Python-style).

Currently, `xonsh.xontribs.xontrib_context()` handles this.

**Comment by astronouth7303** [comment]
The other, more pragmatic, reason for the change is that it's currently really easy for xontribs to pollute the shell context by following standard Python coding conventions. Specifying it this way makes behavior clearer to Python developers, and gives xontributors a tool to directly control what gets exported.

**Comment by scopatz** [comment]
Thanks for opening this up, @astronouth7303!  I think you more pragmatic proposal for `__all__` is something that should be discussed in a separate issue.  At least @gforsyth and I seem in favor of it.

As for this broader proposal, I am all for simplifying the code base.  However, I am not sure that there is much we can do here.  Each of these are implemented in slightly different ways because they have slightly different use cases.  That said, I am in favor of finding whatever commonality is there that hasn't already been implemented.  Ok, let's dive in.

* Python `import`s, as you said, should remain unchanged.
* sourcing should be like typing the file directly into the shell and executing it.  This has the important feature that if you source the same file twice in a row it should be fully executed twice in a row.  This is distinct from importing a file twice, which will only execute it once. 

I think we are all on board so far.

However, with xonshrc files the situation is a little different.  These are like sourcing the file in the sense that its contents should be brought into the local context.  However, it is also like importing int the sense that xonshrc files are meant to only be executed once.  This has important differences implementation-wise.  Sourced files are discovered and [directly fed to the execer](https://github.com/xonsh/xonsh/blob/master/xonsh/aliases.py#L249), whereas xonshrc files are able to be [pickled/marshaled](https://github.com/xonsh/xonsh/blob/master/xonsh/environ.py#L1442) and don't need to be recompiled each time. This has important performance implications for start up.

Moving, on xontrib loading, like you said, is the most like regular Python importing. However, I believe that the `__all__` proposal is actually proposing new behaviour for something that is undefined in Python.  Specifically, `__all__` is only for packages to specifiy which modules from that package they should import: https://docs.python.org/3/tutorial/modules.html#importing-from-a-package

This proposal is to expand `__all__` to also mean to grab members from modules and import them. I don't think this is a huge deal because at least CPython will use `__all__` in modules in this way, even though it is undefined. Since xontribs and xonshrc files are all implemented as modules (to date), I thought this distinction was worth pointing out.

**Comment by scopatz** [comment]
tl;dr I don't have a problem with the `__all__` stuff, but I don't know that we can simplify the various execution pathways more than they are.

**Comment by astronouth7303** [comment]
For reference, the [Python language reference](https://docs.python.org/3/reference/simple_stmts.html#from) says:

> The public names defined by a module are determined by checking the module’s namespace for a variable named `__all__`; if defined, it must be a sequence of strings which are names defined or imported by that module. The names given in `__all__` are all considered public and are required to exist. If `__all__` is not defined, the set of public names includes all names found in the module’s namespace which do not begin with an underscore character (`'_'`). `__all__` should contain the entire public API. It is intended to avoid accidentally exporting items that are not part of the API (such as library modules which were imported and used within the module).

**Comment by laerus** [comment]
hey @astronouth7303 think we can close this?
